pkg/virt-api/webhooks/mutating-webhook/mutators/vm-mutator.go

```diff
@@ -96,19 +98,39 @@ func (mutator *VMsMutator) Mutate(ar *admissionv1.AdmissionReview) *admissionv1.
 }
 
 func (mutator *VMsMutator) setDefaultMachineType(vm *v1.VirtualMachine) {
+	// Nothing to do, let's the validating webhook fail later
 	if vm.Spec.Template == nil {
-		// nothing to do, let's the validating webhook fail later
 		return
 	}
-	machineType := mutator.ClusterConfig.GetMachineType()
 
-	if machine := vm.Spec.Template.Spec.Domain.Machine; machine != nil {
-		if machine.Type == "" {
-			machine.Type = machineType
-		}
-	} else {
-		vm.Spec.Template.Spec.Domain.Machine = &v1.Machine{Type: machineType}
+	if machine := vm.Spec.Template.Spec.Domain.Machine; machine != nil && machine.Type != "" {
+		return
+	}
+
+	if vm.Spec.Template.Spec.Domain.Machine == nil {
+		vm.Spec.Template.Spec.Domain.Machine = &v1.Machine{}
+	}
+
+	if vm.Spec.Preference != nil {
+		vm.Spec.Template.Spec.Domain.Machine.Type = mutator.setPreferenceMachineType(vm)
+	}
+
+	// Only use the cluster default if the user hasn't provided a machine type or referenced a preference with PreferredMachineType
+	if vm.Spec.Template.Spec.Domain.Machine.Type == "" {
+		vm.Spec.Template.Spec.Domain.Machine.Type = mutator.ClusterConfig.GetMachineType()
+	}
+}
+
+func (mutator *VMsMutator) setPreferenceMachineType(vm *v1.VirtualMachine) string {
+	preferenceSpec, err := mutator.InstancetypeMethods.FindPreferenceSpec(vm)
+	if preferenceSpec != nil && preferenceSpec.Machine != nil {
+		return preferenceSpec.Machine.PreferredMachineType
+	}
+	// Log but ultimately swallow any preference lookup errors here and let the validating webhook handle them
+	if err != nil {
```

Comment:
Error checking is usually done before using the first return value. Can you move this `if` directly below the call to `mutator.InstancetypeMethods.FindPreferenceSpec(vm)` ?

Reply:
Yeah apologies I was just trying to avoid multiple returns when one would do but I take your point, this doesn't look correct.